### PR TITLE
Implement custom tld

### PR DESCRIPTION
**Pull request by nisanth074**
This pull makes the top level domain `.dev` configurable https://github.com/code-mancers/invoker/issues/101

Tasks:

- [x] OS X Support
- [x] Linux Support
- [x] Add tests. Refactor if necessary.
- [x] Test the feature manually once

### CODE REVIEW

**Comment by gnufied** [comment]
Looking good, let me know when it is done though - I will double check.

File: lib/invoker/power/url_rewriter.rb [hunk 1]
```diff
@@ -27,6 +28,12 @@ def extract_host_from_domain(complete_path)
 
       private
 
+      def dev_match_regex
+        # [/([\w.-]+)\.dev(\:\d+)?$/, /([\w-]+)\.dev(\:\d+)?$/]
+        tld = Invoker.tld
+        [/([\w.-]+)\.#{tld}(\:\d+)?$/, /([\w-]+)\.#{tld}(\:\d+)?$/]
```
Comment (kgrz) [inline lib/invoker/power/url_rewriter.rb]: What happens when user configures tld as `.development` vs `development`, for example? is there any check preventing them to check for the starting `.`?

Comment (nisanth074) [inline lib/invoker/power/url_rewriter.rb]: @kgrz I'll be adding a validation for tld. Probably something like
```ruby
class Cli < Thor
  def setup
    validate_tld(options[:tld]) if options[:tld]
    # ...
  end

  private

  def validate_tld(tld)
    unless /^[a-z]+$/ =~ tld
      raise InvalidTldError.new('TLD can only contain lower case alphabets')
    end
  end
end
```
This pull is currently a work in progress. I'll tag you or Hemanth for a review once I'm done.

**Comment by gnufied** [comment]
@nisanth074 any news on this one? Looking forward to getting this sorted.

**Comment by nisanth074** [comment]
@gnufied Apologies for the delay here. I haven't been able to spend much time on invoker lately. I'll be working on this today.

**Comment by gnufied** [comment]
Awesome work so far @nisanth074 . let me know when it is done. I am looking forward to make a new release with this change and other fixes we have made.

**Comment by nisanth074** [comment]
@gnufied Can you review? Here's a summary of the changes I've made

- To configure invoker to use a different top level domain, a user can specify the `--tld` flag when running the setup command
  ```
  invoker setup --tld local
  ```
  If a user configures invoker to use a different top level domain, the top level domain is stored in the power config file (~/.invoker/config)

  ```
  ---
  :http_port: 23401
  :https_port: 23402
  :tld: local
  ```
- Changed the ubuntu installer to restart `dnsmasq` after creating the resolver file

I've manually tested the feature in OS X and Ubuntu

**Comment by swrobel** [comment]
:clap:

**Comment by gnufied** [comment]
The part of the code is needlessly complicated. For example, to get `tld` we have 3 sources of truth. `Power::Config`  and `Invoker::Parser::Config` and `Invoker::Power::Setup::Common`. 

This tld thing has been spread all over the place, where as I feel it should be just isolated to `Power::Config` and `Invoker::Power::Setup` classes.

File: lib/invoker/power/power.rb [hunk 2]
```diff
@@ -1,3 +1,31 @@
...
+          tld_value = Invoker.config.tld
+        end
+        Tld.new(tld_value)
+      end
+
+      def tld_value
+        tld.value
+      end
+    end
+  end
```
Comment (gnufied) [inline lib/invoker/power/power.rb]: This is really unnecessary. Just have `tld` is a small string. Just store it as a part of `Setup` class ?

File: lib/invoker/power/setup/common.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+          Invoker::Power.tld
+        end
+
+        def tld_value
+          Invoker::Power.tld_value
+        end
+      end
+    end
+  end
+end
```
Comment (gnufied) [inline lib/invoker/power/setup/common.rb]: I have no idea why this file exists.

File: lib/invoker/power/setup/distro/ubuntu.rb [hunk 4]
```diff
@@ -4,6 +4,9 @@ module Invoker
   module Power
     module Distro
       class Ubuntu < Debian
+        def restart_services
+          system('service dnsmasq restart')
+        end
```
Comment (gnufied) [inline lib/invoker/power/setup/distro/ubuntu.rb]: Why TLD change includes this stuff? This is old upstart way of starting services. All new Ubuntu distros ship with systemd and I thought we moved to systemd for this stuff.

File: lib/invoker/power/tld.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    class Tld
+      def self.new(tld_value = nil)
+        tld_value.nil? ? DefaultTld.new : CustomTld.new(tld_value)
+      end
+
+      def self.default_value
+        DefaultTld.value
+      end
+    end
+  end
```
Comment (gnufied) [inline lib/invoker/power/tld.rb]: Again this file is completely unnecessary.

File: lib/invoker/power/dns.rb [hunk 6]
```diff
@@ -31,7 +31,7 @@ def resource_class_matches?(resource_class)
       end
 
       def name_matches?(name)
-        name =~ /.*\.dev/
+        name =~ /.*\.#{Invoker::Power.tld_value}/
```
Comment (gnufied) [inline lib/invoker/power/dns.rb]: So we really have 3 places from which we can fetch tld value. Why not just use `Invoker.config.tld` ?

**Comment by gnufied** [comment]
Another thing is - we will have to probably provide a migration path for existing users. @nisanth074 This PR looks great tbh. I am happy to merge this if we can sort following:

1. Provide a migration path for existing users. i.e - what if I am running invoker and want to change my tld?

2. This is rather my subjective opinion and let me know if I am wrong, but part of me feels there are too many sources of truth for tld. In other words, we read currently derived tld value from too many places and it is not clear which one is right one.We should probably try to standardize on this.

File: lib/invoker/power/setup/common.rb (hunk 3 above)
Comment (nisanth074) [inline lib/invoker/power/setup/common.rb]: This is an old file from a work in progress commit that should have been removed. Doing it now.

File: lib/invoker/power/setup/distro/ubuntu.rb (hunk 4 above)
Comment (nisanth074) [inline lib/invoker/power/setup/distro/ubuntu.rb]: I was testing on Ubuntu 14.04 LTS which uses upstart. I had an issue where dnsmasq wasn't being restarted during invoker setup. Didn't realize that Ubuntu have moved to systemd now. Will revert this change.

File: lib/invoker/power/power.rb (hunk 2 above)
Comment (nisanth074) [inline lib/invoker/power/power.rb]: `tld` is also used in the `UrlRewriter` class https://github.com/nisanth074/invoker/tree/784916d1e643128ccaa9332472e8133daa67db9f/lib/invoker/power. So, I decided to not store `tld` in the `Setup` class and instead store it in the `Power` class.

I don't mind having `tld` as a simple string (instead of an object) if it makes the code more readable. Will change it.

**Comment by nisanth074** [comment]
@gnufied Thanks for the review and the feedback.

How often are users likely to do this? Is it worth working on? Currently, a user can change the tld by uninstalling invoker and then reinstalling it.
```
sudo invoker uninstall
sudo invoker setup --tld my
```
> Provide a migration path for existing users. i.e - what if I am running invoker and want to change my tld?

I completely agree. Tld value was being derived from too many places. I'm changing the pull to fetch the tld value using `Invoker.config.tld`. During setup, since `Invoker.config` is unavailable, I'll be fetching the tld value from `Invoker::Power::Setup.tld`.
> This is rather my subjective opinion and let me know if I am wrong, but part of me feels there are too many sources of truth for tld. In other words, we read currently derived tld value from too many places and it is not clear which one is right one.We should probably try to standardize on this.

**Comment by nisanth074** [comment]
@gnufied Also, should we rename the command line flag as `--domain`
```
invoker setup --domain local
```
`--tld` seems a bit cryptic

**Comment by nisanth074** [comment]
@gnufied Done here. Should we work on domain migration? Is it a frequent usecase? Also, any thoughts on renaming the command line flag as `--domain`?

File: lib/invoker/power/setup/osx_setup.rb [hunk 7]
```diff
@@ -1,10 +1,39 @@
 module Invoker
   module Power
     class OsxSetup < Setup
-      RESOLVER_FILE = "/etc/resolver/dev"
-      RESOLVER_DIR = "/etc/resolver"
       FIREWALL_PLIST_FILE = "/Library/LaunchDaemons/com.codemancers.invoker.firewall.plist"
 
+      class << self
+        # @!group Helpers for use in tests
```
Comment (gnufied) [inline lib/invoker/power/setup/osx_setup.rb]: Did we just add these for tests? If so - that seems backward. Also, `install` method in setup class is being called via an instance not via class method.

I specifically don't like class vs instance method confusion and hence you will notice all existing methods are instance methods except `Setup.install` and `Setup.uninstall`, but they themselves instantiate right installer classes and call instance on them.

What I mean is - lets not have this confusion about class & instance methods and let everything be instance methods. Test has to work with this design and this code shouldn't be modified so as tests can work.

File: lib/invoker/power/setup/osx_setup.rb [hunk 8]
```diff
@@ -1,10 +1,39 @@
...
+        attr_writer :resolver_file_name
+
+        def resolver_dir
+          return @resolver_dir if @resolver_dir
+          '/etc/resolver'
+        end
+
+        def resolver_file_name
+          return @resolver_file_name if @resolver_file_name
+          Invoker::Power::Setup.tld
```
Comment (gnufied) [inline lib/invoker/power/setup/osx_setup.rb]: Why does this return `tld` ?

File: lib/invoker/power/setup.rb [hunk 9]
```diff
@@ -1,15 +1,25 @@
...
     class Setup
       attr_accessor :port_finder
-      def self.install
+
+      def self.install(options = {})
+        if options[:tld]
+          validate_tld(options[:tld])
+          Invoker::Power::Setup.tld = options[:tld]
+        end
+
```
Comment (gnufied) [inline lib/invoker/power/setup.rb]: this is how I would do:

```
`selected_installedr_class = installer_klass
installer = selected_installer_class.new(tld)
installer.install
```

So as `tld` itself becomes an attribute of installer class.
